### PR TITLE
Conversation handler

### PR DESCRIPTION
**Pull request by karabutov**

### CODE REVIEW

File: bot/bot.py [hunk 1]
```diff
@@ -1,19 +1,15 @@
...
-    CallbackContext,
     CommandHandler,
     Filters,
     MessageHandler,
     Updater,
-    CallbackQueryHandler,
 )
-from typing import NoReturn
 
+import handlers
```
Comment (alexbuchkin) [inline bot/bot.py]: Absolute import required

File: bot/buttons.py [hunk 2]
```diff
@@ -1,10 +1,63 @@
+from typing import List
+
 from telegram import InlineKeyboardButton
 from telegram import InlineKeyboardMarkup
+from telegram import ReplyKeyboardMarkup
+from telegram import KeyboardButton
+import constants
```
Comment (alexbuchkin) [inline bot/buttons.py]: Absolute import required

File: bot/buttons.py [hunk 3]
```diff
@@ -1,10 +1,63 @@
...
+
+from database.new_types import User, Event
```
Comment (alexbuchkin) [inline bot/buttons.py]: Let's rename **new_types.py** to **types.py**

File: bot/buttons.py (hunk 2 above)
Comment (alexbuchkin) [inline bot/buttons.py]: Please replace **constants** with a bit more meaningful name

File: bot/buttons.py [hunk 4]
```diff
@@ -1,10 +1,63 @@
...
+    return keyboard
+
+
+def get_menu_keyboard() -> InlineKeyboardMarkup:
+    select_event = InlineKeyboardButton('Выбрать мероприятие', callback_data=constants.SELECT_EVENT)
+    join_event = InlineKeyboardButton('Присоединиться к меропиятию', callback_data=constants.JOIN_EVENT)
+    create_event = InlineKeyboardButton('Создать мероприятие', callback_data=constants.CREATE_EVENT)
+
+    keyboard = InlineKeyboardMarkup([[select_event], [join_event], [create_event]])
+    return keyboard
```
Comment (alexbuchkin) [inline bot/buttons.py]: Pls delete one of them, must be a typo

File: bot/buttons.py [hunk 5]
```diff
@@ -1,10 +1,63 @@
...
+
+
+def get_empty_keyboard():
```
Comment (alexbuchkin) [inline bot/buttons.py]: Add return value type

File: bot/buttons.py [hunk 6]
```diff
@@ -1,10 +1,63 @@
...
+    return InlineKeyboardMarkup([])
+
+
+def get_event_commands_keyboard () -> InlineKeyboardMarkup:
```
Comment (alexbuchkin) [inline bot/buttons.py]: **get_event_commands_keyboard ()** - delete space

File: bot/constants.py [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#COMANDS
```
Comment (alexbuchkin) [inline bot/constants.py]: Typo

Comment (alexbuchkin) [inline bot/constants.py]: There must be a single space between # and the beginning of the comment, see https://www.python.org/dev/peps/pep-0008/#inline-comments

File: bot/constants.py [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+SELECT_EVENT = '\\select_event'
```
Comment (alexbuchkin) [inline bot/constants.py]: Maybe you meant **'/select_event'**?

File: bot/handlers.py [hunk 9]
```diff
@@ -0,0 +1,541 @@
...
+    CallbackContext,
+    CommandHandler,
+    Filters,
+    MessageHandler,
+    CallbackQueryHandler,
+    ConversationHandler,
+)
+
+import buttons
+import constants
```
Comment (alexbuchkin) [inline bot/handlers.py]: Absolute import required. We need to read [pep8 import instructions](https://www.python.org/dev/peps/pep-0008/#imports) together

File: bot/handlers.py [hunk 10]
```diff
@@ -0,0 +1,541 @@
...
+
+from database.new_types import (
+    User,
+)
+
+_CURRENT_EVENT_TOKEN = 'CURRENT_EVENT_TOKEN'
+_EXPENSE = 'EXPENSE'
+_DEBT = 'DEBT'
+
+_EVENT_NAME_STATE = 4
```
Comment (alexbuchkin) [inline bot/handlers.py]: Is it really necessary to use integers instead of enums?

File: bot/handlers.py [hunk 11]
```diff
@@ -0,0 +1,541 @@
...
+        update: Update,
+        _: CallbackContext,
+    ) -> NoReturn:
+        id_ = update.effective_user.id
+        name = update.effective_user.username
+
+        if self._splitwise.user_exists(id_):
+            name = self._splitwise.get_user_info(id_).name
+            update.effective_chat.send_message(f'Привет, {name}! Ты уже был(а) здесь!')
+            update.effective_chat.send_message('Меню:', reply_markup=buttons.get_menu_keyboard())
```
Comment (alexbuchkin) [inline bot/handlers.py]: Move this after if-statement

**Review by alexbuchkin — CHANGES_REQUESTED** [review]
pls read this

File: app/splitwise.py [hunk 12]
```diff
@@ -1,193 +1,205 @@
...
+        users = self.get_users_of_event(event_token)
+        usernames = {user.id: user.name for user in users}
+        users_balance = defaultdict(lambda: {
+            'lent': 0,
+            'owed': 0,
+        })
+
+        event_expenses = self.conn.get_event_expenses(event_token)
+        for expense in event_expenses:
+            users_balance[expense.lender_id]['lent'] += expense.sum
```
Comment (alexbuchkin) [inline app/splitwise.py]: We have to be sure that expense amount is equal to sum of all debts related to the current expense

File: bot/__main__.py [hunk 13]
```diff
@@ -1,6 +1,6 @@
 import os
 
-from bot.bot import TelegramBot
+from bot import TelegramBot
```
Comment (alexbuchkin) [inline bot/__main__.py]: Pls use absolute import

File: bot/bot.py [hunk 14]
```diff
@@ -1,628 +1,34 @@
...
         dispatcher = self.updater.dispatcher
-        dispatcher.add_handler(CommandHandler("create_event", self.create_event_handler))
-        dispatcher.add_handler(CommandHandler("add_expense", self.add_expense_handler))
-        dispatcher.add_handler(CommandHandler("show_debts", self.show_debts_handler))
-        dispatcher.add_handler(CommandHandler("users_of_event", self.users_of_event_handler))
-        dispatcher.add_handler(CommandHandler("join_event", self.join_event_handler))
-        dispatcher.add_handler(CommandHandler("start", self.start_handler))
-        dispatcher.add_handler(CallbackQueryHandler(self.callback_query_handler))
-        dispatcher.add_handler(MessageHandler(Filters.text, self.text_handler))
+        dispatcher.add_handler(handlers.MenuButtonsConversationHandler().get_conversation_handler())
```
Comment (alexbuchkin) [inline bot/bot.py]: Does order of adding handlers matter here?

File: bot/buttons.py [hunk 15]
```diff
@@ -1,10 +1,63 @@
+from typing import List
+
 from telegram import InlineKeyboardButton
 from telegram import InlineKeyboardMarkup
+from telegram import ReplyKeyboardMarkup
+from telegram import KeyboardButton
```
Comment (alexbuchkin) [inline bot/buttons.py]: `from telegram import (
    InlineKeyboardButton,
    InlineKeyboardMarkup,
    ReplyKeyboardMarkup,
    KeyboardButton,
)`

File: bot/buttons.py [hunk 16]
```diff
@@ -1,10 +1,63 @@
...
+    keyboard = [
+        [InlineKeyboardButton(event.name, callback_data=event.token)]
+        for event in events
+    ]
+    keyboard.append([InlineKeyboardButton('Назад', callback_data=constants.CANCEL)])
     return InlineKeyboardMarkup(keyboard)
+
+
+def get_cancel_button() -> InlineKeyboardMarkup:
+    keyboard = [[InlineKeyboardButton('Назад', callback_data='cancel')]]
```
Comment (alexbuchkin) [inline bot/buttons.py]: Why not `constants.CANCEL`?

File: bot/buttons.py [hunk 17]
```diff
@@ -1,10 +1,63 @@
...
+
+from database.new_types import User, Event
+
+
+def get_user_buttons(users: List[User]) -> InlineKeyboardMarkup:
+    keyboard = [
+        [InlineKeyboardButton(user.name, callback_data=user.id)]
+        for user in users
+    ]
+    keyboard.append([InlineKeyboardButton("Закончить", callback_data=constants.CANCEL)])
```
Comment (alexbuchkin) [inline bot/buttons.py]: Use single quotes here

File: bot/handlers.py [hunk 18]
```diff
@@ -0,0 +1,552 @@
...
+            return _ASKING_FOR_ACTION
+
+    def create_event(
+        self,
+        update: Update,
+        _: CallbackContext,
+    ) -> int:
+        update.effective_chat.send_message('Введи название мероприятия или нажмите кнопку \'Отмена\'',
+                                           reply_markup=buttons.get_cancel_button())
+        return _EVENT_NAME_STATE
```
Comment (alexbuchkin) [inline bot/handlers.py]: Why there is a `_` prefix? Let's get rid of it

File: bot/handlers.py [hunk 19]
```diff
@@ -0,0 +1,552 @@
...
+        events = self._splitwise.get_user_events(user_id)
+        update.effective_chat.send_message('Выбери мероприятие', reply_markup=buttons.get_event_buttons(events))
+        return _ASKING_FOR_ACTION
+
+    def fallback_handler(
+        self,
+        update: Update,
+        _: CallbackContext,
+    ):
+        update.effective_chat.send_message("Ошибка, но мы не знаем что произошло")
```
Comment (alexbuchkin) [inline bot/handlers.py]: Use single quotes here

File: bot/handlers.py [hunk 20]
```diff
@@ -0,0 +1,552 @@
...
+        update: Update,
+        _: CallbackContext,
+    ):
+        if update.callback_query.data == constants.CANCEL:
+            update.callback_query.answer('Создание мероприятия отменено')
+            update.callback_query.edit_message_text('Меню:')
+            update.callback_query.edit_message_reply_markup(reply_markup=buttons.get_menu_keyboard())
+            return _END
+        else:
+            update.callback_query.answer('Не на ту кнопку жмешь')
```
Comment (alexbuchkin) [inline bot/handlers.py:197]: It was a bit rude, mate!

File: bot/handlers.py [hunk 21]
```diff
@@ -0,0 +1,552 @@
...
+            update.callback_query.edit_message_reply_markup(reply_markup=buttons.get_menu_keyboard())
+            return _END
+        else:
+            update.callback_query.answer('Не на ту кнопку жмешь',)
+            return None
+
+    def fallbacks_handler(
+            self,
+            update: Update,
+            _: CallbackContext,
```
Comment (alexbuchkin) [inline bot/handlers.py]: Too big indent, must be 4 spaces

File: bot/handlers.py [hunk 22]
```diff
@@ -0,0 +1,552 @@
...
+
+    def expense_sum(
+        self,
+        update: Update,
+        context: CallbackContext,
+    ) -> Optional[int]:
+        # let's use integer division and forget about cents for some time
+        expense_sum = update.effective_message.text
+        try:
+            expense_sum = int(expense_sum)
```
Comment (alexbuchkin) [inline bot/handlers.py]: Add validation here

File: database/__main__.py [hunk 23]
```diff
@@ -3,29 +3,30 @@
 
 if __name__ == '__main__':
```
Comment (alexbuchkin) [inline database/__main__.py:4]: Let's remove whole file

File: database/connector.py [hunk 24]
```diff
@@ -1,207 +1,239 @@
...
-
-                FOREIGN KEY (expense_id) REFERENCES expenses(id),
-                FOREIGN KEY (lender_id)  REFERENCES users(id),
-                FOREIGN KEY (debtor_id)  REFERENCES users(id)
-            )
-        ''')
+        current_dir = Path(__file__).resolve().parent
+        try:
+            with open(current_dir.joinpath('sql/create_tables.sql'), 'r') as file:
+                cursor.executescript(file.read())
```
Comment (alexbuchkin) [inline database/connector.py:38]: Replace this with [importlib-resources](https://importlib-resources.readthedocs.io/en/latest/using.html#example)

**Review by alexbuchkin — CHANGES_REQUESTED** [review]

File: bot/bot.py (hunk 14 above)
Comment (alexbuchkin) [reply bot/bot.py]: https://python-telegram-bot.readthedocs.io/en/stable/telegram.ext.dispatcher.html#telegram.ext.Dispatcher.add_handler
TL;DR: Order and priority counts

Comment (alexbuchkin) [reply bot/bot.py]: Order is right, nothing to fix

File: bot/handlers.py (hunk 20 above)
Comment (alexbuchkin) [reply bot/handlers.py:197]: Tech debt

File: database/__main__.py (hunk 23 above)
Comment (alexbuchkin) [reply database/__main__.py:4]: Tech debt

File: database/connector.py (hunk 24 above)
Comment (alexbuchkin) [reply database/connector.py:38]: Tech debt
Also we should not use importlib-resources, because I'm scared

**Review by alexbuchkin — COMMENTED** [review]
Everything fixed

**Review by alexbuchkin — APPROVED** [review]
